### PR TITLE
Skeleton-health-calculations-and-testing

### PR DESCRIPTION
**Pull request by L00TG0BLIN**
Damage is used for testing/showing the outputs of damage to specific body parts
Thresholds is used to store threshold values for each entity so that they won't have to be reinitialized with every new entity that spawns
Health is used to calculate damage types to a specific body part
To be merged with skeleton/detaching

### CODE REVIEW

File: Assets/Scripts/Player/Damage.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    // Start is called before the first frame update
+    void Start()
+    {
+    }
+
+    void Update()
+    {
+        if (Input.GetKeyDown(KeyCode.Q) && currentTarget > 0)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Damage.cs]: These code blocks should be condensed into a single if() for each keycode followed by a nested if() for the second parameters, to reduce the amount of times GetKeyDown() is called.

File: Assets/Scripts/Player/Health.cs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+    }
+
+    private limbStatus[] statusBools = new limbStatus[7];
+    private HealthInfo[] limbHealthInfo = new HealthInfo[7];
+
+    //enum for entity types
+    public enum Entities
+    {
+        Organic,
+        Onebi,
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: ?

File: Assets/Scripts/Player/Health.cs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        while(numBleedingLimbs > 0)
+        {
+            currentBloodVol -= numBleedingLimbs * 10;
+            //Debug.Log("Current Blood Volume: " + currentBloodVol);
+            yield return new WaitForSecondsRealtime(5);
+        }
+    }
+
+    //Coroutine to regain blood overtime. Currently 2 cc's of blood are gained every 7 seconds
+    private IEnumerator BloodRegen()
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: This function should not require a while loop within as long as you check requirements before calling it within whatever life loop you end up using.
i.e.
```
private void life()
{
    if(currentBloodVol < maxBloodVol)
    {
        handleBloodRegen()
    }
}

private IEnumerator handleBloodRegen()
{
    currentBloodVol = Min(currentBloodVol + bloodRegenAmount, maxBloodVol)
}
````

File: Assets/Scripts/Player/Health.cs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        }
+        if(crit == false && currentHealth <= -50)
+        {
+            scrit = false;
+            crit = true;
+        }
+    }
+
+    //Coroutine to lose blood overtime. Currently (10 * the number of bleeding limbs) cc's of blood are lost every 5 seconds
+    private IEnumerator BloodLoss()
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: This function should not require a while loop within as long as you check requirements before calling it within whatever life loop you end up using.
i.e.
```
private void life()
{
    if(numBleedingLimbs)
    {
        handleBloodLoss()
    }
}

private IEnumerator handleBloodLoss()
{
    currentBloodVol = Max(currentBloodVol - (numBleedingLimbs * 10), 0)
}
````

File: Assets/Scripts/Player/Health.cs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+            //Calculates if a limb is blistered
+            if (statusBools[target].blistered == false && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minBlistered, Thresholds.thresholds[(int)Entities.Organic].maxBlistered) * limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].burn)
+            {
+                statusBools[target].blistered = true;
+            }
+
+        }
+
+        //Determines if the player is in soft crit or regular crit
+        if(scrit == false && currentHealth <= 0)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]:
```
if(currentHealth <= DAMAGE_THRESHOLD_CRIT)
{
    critLevel = CRIT_LEVEL_HARD
}
else if(currentHealth <= DAMAGE_THRESHOLD_SOFTCRIT)
{
    critLevel = CRIT_LEVEL_SOFT
}
````
You don't need two variables for crit when you can store them as 0 1 2

File: Assets/Scripts/Player/Health.cs [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        if (dmgType == 0)
+        {
+            //Calculates if a limb is bruised
+            if (statusBools[target].bruised == false && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minBruise, Thresholds.thresholds[(int)Entities.Organic].maxBruise)* limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].brute)
+            {
+                statusBools[target].bruised = true;
+            }
+
+            //Calculates if a limb is bleeding. If true, begin the BloodLoss Coroutine
+            if (statusBools[target].bleeding == false && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minBleed, Thresholds.thresholds[(int)Entities.Organic].maxBleed) * limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].brute)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: bleeding statuses shouldn't be determined by how much damage is taken, rather the type of damage taken (piercing/blunt/slashing)

File: Assets/Scripts/Player/Health.cs [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        dmgDebug();
+       }
+
+    public void dmgEval(int target, int dmgType)
+    {
+        //Determines the status of an affected limb based on the brute damage
+        if (dmgType == 0)
+        {
+            //Calculates if a limb is bruised
+            if (statusBools[target].bruised == false && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minBruise, Thresholds.thresholds[(int)Entities.Organic].maxBruise)* limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].brute)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: bruising statuses shouldn't be determined by how much damage is taken, rather the type of damage taken (piercing/blunt/slashing)

File: Assets/Scripts/Player/Health.cs [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+            //Calculates if a limb is crippled
+            if (statusBools[target].crippled == false && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minCrippled, Thresholds.thresholds[(int)Entities.Organic].maxCrippled) * limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].brute)
+            {
+                statusBools[target].crippled = true;
+            }
+        }
+
+        //Determines the status of an affected limb based on the burn damage
+        else if (dmgType == 1)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: do not use magical numbers

File: Assets/Scripts/Player/Health.cs [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            limbHealthInfo[(int)SkeletalSections.overall].suffocation += dmg;
+
+        dmgEval(target, dmgType);
+        dmgDebug();
+       }
+
+    public void dmgEval(int target, int dmgType)
+    {
+        //Determines the status of an affected limb based on the brute damage
+        if (dmgType == 0)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: do not use magical numbers

File: Assets/Scripts/Player/Health.cs [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+        limbHealthInfo[(int)SkeletalSections.leftLeg].maxLimbHP = limbHealthInfo[(int)SkeletalSections.rightLeg].maxLimbHP = maxHealth;
+
+        //Initilizes Thresholds dependent on Entity type
+        Thresholds.InitializeThresholds();
+    }
+
+    //Updates the damage value of corresponding damage to corresponding body part targetted
+    public void dmgUpdate(int target, int dmgType , float dmg)
+    {
+        if (dmgType == 0)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: do not use magical numbers.

File: Assets/Scripts/Player/Health.cs [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+        {
+            limbHealthInfo[target].brute += dmg;
+            limbHealthInfo[(int)SkeletalSections.overall].brute += dmg;
+        }
+        else if (dmgType == 1)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: This should be a switch statement

File: Assets/Scripts/Player/Health.cs [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+        limbHealthInfo[(int)SkeletalSections.torso].maxLimbHP = limbHealthInfo[(int)SkeletalSections.head].maxLimbHP =
+        limbHealthInfo[(int)SkeletalSections.leftArm].maxLimbHP = limbHealthInfo[(int)SkeletalSections.rightArm].maxLimbHP =
+        limbHealthInfo[(int)SkeletalSections.leftLeg].maxLimbHP = limbHealthInfo[(int)SkeletalSections.rightLeg].maxLimbHP = maxHealth;
+
+        //Initilizes Thresholds dependent on Entity type
+        Thresholds.InitializeThresholds();
+    }
+
+    //Updates the damage value of corresponding damage to corresponding body part targetted
+    public void dmgUpdate(int target, int dmgType , float dmg)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: random space before ,

File: Assets/Scripts/Player/Health.cs [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        //Intializes overall player health
+        currentHealth = maxHealth;
+        currentBloodVol = maxBloodVol;
+
+        //Initilize the HP of each respective limb
+        limbHealthInfo[(int)SkeletalSections.torso].maxLimbHP = limbHealthInfo[(int)SkeletalSections.head].maxLimbHP =
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: you should define an enum array for starting limb healths

File: Assets/Scripts/Player/Damage.cs [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+    public enum DamageType
+    {
+        brute,
+        burn,
+        toxic,
+        suffocation
+    }
+
+    enum SectionTarget
```
Comment (CakeQ) [inline Assets/Scripts/Player/Damage.cs]: you've defined this but haven't used it anywhere

File: Assets/Scripts/Player/Damage.cs [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+            currentTarget++;
+            Debug.Log("Traget ID: " + currentTarget);
+        }
+        else if (Input.GetKeyDown(KeyCode.E) && currentTarget == 5)
+        {
+            currentTarget = 0;
+            Debug.Log("Traget ID: " + currentTarget);
+        }
+
+        if (Input.GetKeyDown(KeyCode.A) && currentDmgType > 0)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Damage.cs]: currentDmgType should not be using magical numbers

File: Assets/Scripts/Player/Health.cs [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+            {
+                statusBools[target].crippled = true;
+            }
+        }
+
+        //Determines the status of an affected limb based on the burn damage
+        else if (dmgType == 1)
+        {
+            //Calculates if a limb is burned
+            if (statusBools[target].burned == false && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minBurn, Thresholds.thresholds[(int)Entities.Organic].maxBurn) * limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].burn)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: if(!statusBools[target].burned

File: Assets/Scripts/Player/Health.cs [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+            {
+                statusBools[target].burned = true;
+            }
+
+            //Calculates if a limb is numb
+            if (statusBools[target].numb == false && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minNumb, Thresholds.thresholds[(int)Entities.Organic].maxNumb) * limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].burn)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: if(!statusBools[target].numb

File: Assets/Scripts/Player/Health.cs [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+            {
+                statusBools[target].numb = true;
+            }
+
+            //Calculates if a limb is blistered
+            if (statusBools[target].blistered == false && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minBlistered, Thresholds.thresholds[(int)Entities.Organic].maxBlistered) * limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].burn)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: if(!statusBools[target].blistered

File: Assets/Scripts/Player/Health.cs [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+                statusBools[target].bleeding = true;
+                numBleedingLimbs++;
+                /*if(numBleedingLimbs == 1)
+                StartCoroutine("BloodLoss");
+                StartCoroutine("BloodRegen");
+                */
+            }
+
+            //Calculates if a limb is crippled
+            if (statusBools[target].crippled == false && Random.Range(Thresholds.thresholds[(int)Entities.Organic].minCrippled, Thresholds.thresholds[(int)Entities.Organic].maxCrippled) * limbHealthInfo[target].maxLimbHP < limbHealthInfo[target].brute)
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: if(!statusBools[target].crippled

**Comment by onebigames** [comment]
As discussed on discord, this PR has not been opened for the right reasons

File: Assets/Scripts/Player/Health.cs (hunk 6 above)
Comment (CakeQ) [reply Assets/Scripts/Player/Health.cs]: if(!statusBools[target].bleeding

File: Assets/Scripts/Player/Health.cs (hunk 7 above)
Comment (CakeQ) [reply Assets/Scripts/Player/Health.cs]: if(!statusBools[target].bruised

File: Assets/Scripts/Player/Health.cs [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+
+    public int maxHealth = 100;
+    public int currentHealth;
+
+    //sets the max blood capacity to 560ccs
+    private float maxBloodVol = 560f;
+    private float currentBloodVol;
+
+    //number of limbs that are currently bleeding
+    private int numBleedingLimbs;
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: This should be a list that stores references to the limbs that are bleeding. You can get an integer value from this by using the .len of the list.

File: Assets/Scripts/Player/Health.cs [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+
+    //Booleans for crit and soft crit
+    private bool scrit = false;
```
Comment (CakeQ) [inline Assets/Scripts/Player/Health.cs]: These two could be combined into a bitflag status var

**Pull request closed without merging**